util/progress/progressui/printer.go

```diff
@@ -325,6 +222,252 @@ func (p *textMux) print(t *trace) {
+}
+
+type textPrinter struct{}
+
+func (text *textPrinter) printVtx(p *textMux, t *trace, v *vertex) {
+	if v.Digest != p.current {
+		if p.current != "" {
+			old := t.byDigest[p.current]
+			if old.logsPartial {
+				fmt.Fprintln(p.w, "")
+			}
+			old.logsOffset = 0
+			old.count = 0
+			fmt.Fprintf(p.w, "#%d ...\n", old.index)
+		}
+
+		if p.notFirst {
+			fmt.Fprintln(p.w, "")
+		} else {
+			if p.desc != "" {
+				fmt.Fprintf(p.w, "#0 %s\n\n", p.desc)
+			}
+			p.notFirst = true
+		}
+
+		fmt.Fprintf(p.w, "#%d %s\n", v.index, p.getVtxName(v))
+	}
+
+	isOpenStatus := p.processStatusUpdates(v, func(s *status) {
+		text.printStatusUpdate(p, v, s)
+	})
+
+	for _, w := range v.warnings[v.warningIdx:] {
+		fmt.Fprintf(p.w, "#%d WARN: %s\n", v.index, w.Short)
+		v.warningIdx++
+	}
+
+	for i, l := range v.logs {
+		if i == 0 && v.logsOffset != 0 { // index has already been printed
+			l = l[v.logsOffset:]
+			fmt.Fprintf(p.w, "%s", l)
+		} else {
+			fmt.Fprintf(p.w, "#%d %s", v.index, []byte(l))
+		}
+
+		if i != len(v.logs)-1 || !v.logsPartial {
+			fmt.Fprintln(p.w, "")
+		}
+		if v.logsBuffer == nil {
+			v.logsBuffer = ring.New(logsBufferSize)
+		}
+		v.logsBuffer.Value = l
+		if !v.logsPartial {
+			v.logsBuffer = v.logsBuffer.Next()
+		}
+	}
+
+	if len(v.logs) > 0 {
+		if v.logsPartial {
+			v.logs = v.logs[len(v.logs)-1:]
+			v.logsOffset = len(v.logs[0])
+		} else {
+			v.logs = nil
+			v.logsOffset = 0
+		}
+	}
+
+	p.current = v.Digest
+	if v.isCompleted() && !isOpenStatus {
+		p.current = ""
+		v.count = 0
+
+		if v.logsPartial {
+			fmt.Fprintln(p.w, "")
+		}
+		if v.Error != "" {
+			if strings.HasSuffix(v.Error, context.Canceled.Error()) {
+				fmt.Fprintf(p.w, "#%d CANCELED\n", v.index)
+			} else {
+				fmt.Fprintf(p.w, "#%d ERROR: %s\n", v.index, v.Error)
+			}
+		} else if v.Cached {
+			fmt.Fprintf(p.w, "#%d CACHED\n", v.index)
+		} else {
+			tm := ""
+			var ivals []interval
+			for _, ival := range v.intervals {
+				ivals = append(ivals, ival)
+			}
+			ivals = mergeIntervals(ivals)
+			if len(ivals) > 0 {
+				var dt float64
+				for _, ival := range ivals {
+					dt += ival.duration().Seconds()
+				}
+				tm = fmt.Sprintf(" %.1fs", dt)
+			}
+			fmt.Fprintf(p.w, "#%d DONE%s\n", v.index, tm)
+		}
+	}
+}
+
+func (text *textPrinter) printStatusUpdate(p *textMux, v *vertex, s *status) {
+	var bytes string
+	if s.Total != 0 {
+		bytes = fmt.Sprintf(" %.2f / %.2f", units.Bytes(s.Current), units.Bytes(s.Total))
+	} else if s.Current != 0 {
+		bytes = fmt.Sprintf(" %.2f", units.Bytes(s.Current))
+	}
+	var tm string
+	endTime := s.Timestamp
+	if s.Completed != nil {
+		endTime = *s.Completed
+	}
+	if s.Started != nil {
+		diff := endTime.Sub(*s.Started).Seconds()
+		if diff > 0.01 {
+			tm = fmt.Sprintf(" %.1fs", diff)
+		}
+	}
+	if s.Completed != nil {
+		tm += " done"
+	}
+	fmt.Fprintf(p.w, "#%d %s%s%s\n", v.index, s.ID, bytes, tm)
+}
+
+type jsonPrinter struct {
+	seen map[digest.Digest]struct{}
+	buf  SolveStatus
```

Comment:
I don't quite get the need for this `buf` property. If it just for passing data when calling methods from `printVtx` then it should be passed as function parameter.

Reply:
This is mostly to prevent large numbers of allocations for the slices in the `SolveStatus` struct. Since these slices are short-lived (used only for the marshaling), we don't need to be careful about the lifetime of the contents of the slice. So the printer keeps a slice around and truncates it to a size of zero but keeps its capacity the same to avoid extra allocations while printing.

It's just most simple to keep it here since this is the only section of code that uses this struct.

Reply:
Ah, ok. I'm not sure if this is premature optimization or not, but add a small comment explaining that this is for slice memory reuse.

Maybe it still makes sense to pass `buf` as a parameter to the other methods(turn them into helper functions).